File: src/utils/data/ValuedNode.java
package utils.data;

public class ValuedNode<T extends Comparable<T>> implements Comparable<ValuedNode<T>> {
    private ValuedNode<T> parent;
    private T data;
    private MinMaxQueue<ValuedNode<T>> childs;
    private Cost culminlativeCost;
    
    public ValuedNode(T data) {
        this.data = data;
        childs = new MinMaxQueue<ValuedNode<T>>();
        culminlativeCost = new Cost();
    }
    
    public int compareTo(ValuedNode<T> node) {
        return data.compareTo(node.getData());
    }
    
    public T getData() {
        return data;
    }
    
    public ValuedNode<T> getParent() {
        return parent;
    }
    
    public T getParantData() {
        return (parent != null)? parent.getData() : null;
    }
    
    public void setParent(ValuedNode<T> p) {
        parent = p;
    }
    
    public boolean hasChild() {
        return !childs.isEmpty();
    }
    
    public int getChildSize() {
        return childs.size();
    }
    
    public void addChild(T data) {
        ValuedNode<T> n = new ValuedNode<T>(data);
        n.setParent(this);
        childs.add(n);
    }
    
    public void addChild(ValuedNode<T> n) {
        if (n != null) {
            n.setParent(this);
            childs.add(n);
        }
    }
    
    public ValuedNode<T> getChild(int i) {
        if (i >= 0 && i < childs.size()) {
            return childs.get(i);
        } else {
            return null;
        }
    }
    
    public ValuedNode<T> getMinChild() {
        return childs.getMin();
    }
    
    public ValuedNode<T> getMaxChild() {
        return childs.getMax();
    }
    
    public T getChildData(int i) {
        if (i >= 0 && i < childs.size()) {
            return childs.get(i).getData();
        } else {
            return null;
        }
    }
    
    public Cost getCulminlativeCost() {
        return culminlativeCost;
    }
    
    public class Cost implements Comparable<Cost> {
        private double cost;
        
        public Cost() {
            cost = 0;
        }
        
        public double getCost() {
            return cost;
        }
        
        public void setCost(double c) {
            cost = c;
        }
        
        public void increment(double c) {
            cost += c;
        }
        
        public int compareTo(Cost c) {
            return (int) (cost - c.cost);
        }
    }
}

File: src/utils/data/ValuedNode.kt
package utils.data;

class ValuedNode<T: Comparable<T>>: Comparable<ValuedNode<T>> {
    var parent: ValuedNode<T>?
        private set
    var data: T
        private set
    private val children: MinMaxQueue<ValuedNode<T>>
    
    var culminlativeCost: Cost
        private set
    
    constructor(data: T) {
        this.data = data
        this.parent = null
        children = MinMaxQueue<ValuedNode<T>>()
        culminlativeCost = Cost()
    }
    
    override fun compareTo(other: ValuedNode<T>): Int {
        return data.compareTo(other.data)
    }
    
    fun getParentData(): T? {
        return parent?.data
    }
    
    fun hasChild(): Boolean {
        return children.isNotEmpty()
    }
    
    fun getChildSize(): Int {
        return children.size
    }
    
    fun addChild(data: T) {
        val n = ValuedNode<T>(data)
        n.parent = this
        children.add(n)
    }
    
    fun addChild(node: ValuedNode<T>) {
        node.parent = this
        children.add(node)
    }
    
    fun getChild(index: Int): ValuedNode<T>? {
        if (index >= 0 && index < children.size) {
            return children.get(index)
        } else {
            return null
        }
    }
    
    fun getMinChild(): ValuedNode<T> {
        return children.getMin()
    }
    
    fun getMaxChild(): ValuedNode<T> {
        return children.getMax()
    }
    
    fun getChildData(index: Int): T? {
        if (index >= 0 && index < children.size) {
            return children.get(index).data
        } else {
            return null
        }
    }
    
    class Cost : Comparable<Cost> {
        var cost: Double = 0.0

        fun increment(c: Double) {
            cost += c;
        }
        
        override fun compareTo(other: Cost): Int {
            return (cost - other.cost).toInt()
        }

        init {
            cost = 0.0
        }
    }
}
